lib/Analysis/Allocation.cpp

```diff
@@ -625,12 +622,17 @@ class AllocationAnalysis {
     // TODO(Keren): We are wasting memory here.
     // Nodes with color2 can actually start with 24.
     for (auto x : buffers) {
-      size_t adj = 0;
+      size_t newOffset = 0;
       for (auto y : interference.lookup(x)) {
-        adj = std::max(adj, bufferStart.lookup(y) + y->size);
+        newOffset = std::max(newOffset, y->offset + y->size);
+      }
+      if (colors.lookup(x) != 0) {
```

Comment:
What does `colors.lookup(x) != 0` mean?

Reply:
A non-zero color means there was an interference. Previously it was multiplied by the bump value, so in the case of color == 2, it would jump twice as far past the interference buffer.

Reply:
interference only depends on the size of the `colors`, right? Number of colors means the number of conflict groups.

Reply:
For example, if `num(colors) == 3`, would `color = 1` jump a too small range to fail solving the conflict that it could solve?

Reply:
The newOffset is based of the max offset of the interference buffers, so it shouldn't matter what color it is, it will just move the buffer past the max interference buffer.

Reply:
The purpose of jumping `color[i] * offset` is to leave enough space to solve conflicts.

For example, if you have:

color0 = [0, 10)
color1 = [0, 10)
...
color(n-1) = [0, 10)

color(n-1) will be updated to (n-1) * 10 
The space between 10 to (n-1) * 10 could be reused by other buffers.

I think the problem here is probably that I mistakenly updated `bufferStart[x]` too early.

A better algorithm would be:

```
    for (auto x : buffers) {
      size_t newOffset = 0;
      for (auto y : interference.lookup(x)) {
        newOffset = std::max(newOffset, bufferStart[y]+ y->size);
      }
      if (colors.lookup(x) != 0)
        x->setOffsetAligned(newOffset);
      allocation->sharedMemorySize =
          std::max(allocation->sharedMemorySize, x->offset + x->size);
    }
    for (auto x : buffers) {
        // Update bufferStart here
    }
```

Reply:
This way bufferStart records the previous offset but not the updated offset.
But bufferStart will be updated after the conflict resolving loop

Reply:
> The space between 10 to (n-1) * 10 could be reused by other buffers.

Otherwise, the minimum offset now will be the new offset + size of `color(n-1)`, which is n * 10. And the space between 10 to n * 100 is wasted.

Reply:
Example:
color(0):  buf0 (size=1024)
color(1):  buf1 (size=32)
color(2):  buf2 (size=1024)

The problem with using color * (buf0->offset + buf0->size) for buf2 is that it assumes that buf1 will have the same size as buf0. Even when the newOffset = buf0->offset + color * buf0->size, which is much better if offset != 0, still results in this packing:

buf0 => [0, 1024)
buf1 => [1024, 1056)     --  buf0->offset + color(==1) * buf0->size
buf2 => [2048, 3072)     --  buf0->offset + color(==2) * buf0->size

However, with this PRs approach, the newOffset for buf2 would be max(buf0->offset + buf0->size, buf1->offset + buf1->size) assuming buf1 was already updated. So buf2 would slide to [1056, 2080) assuming alignment == 32.

buf0 => [0, 1024)
buf1 => [1024, 1056)     --  buf0->offset + buf0->size
buf2 => [1056, 2080)     --  buf1->offset + buf1->size

Reply:
It should also converge quicker than subsequent update loop for bufferStart since subsequent buffers will be packed based on new offsets of interference buffers already processed.

Reply:
So I was suggesting that we don't use `buf0->offset`, Instead using the buffer start recorded previously

Reply:
Ah, I should have written it using bufferStart to avoid confusion. But I think we were already agreeing not to use color * size, because of the case I show above?

So if we use bufferStart to defer updates until after the conflict resolution loop, then for the example above, it would bump both buf1 and buf2 to just after buf0, thus needing another iteration to bump buf2 past buf1. But moving slower may allow for filling gaps better when a subsequent buffer bump would trigger another iteration. I should run some experiments but in the meantime I can revert to using bufferStart with a trailing update loop if you prefer.

Reply:
> So if we use bufferStart to defer updates until after the conflict resolution loop, then for the example above, it would bump both buf1 and buf2 to just after buf0, thus needing another iteration to bump buf2 past buf1.

I think there's no optimal allocation algorithm for this problem. I'm OK with the current implementation as long as we understand the pros and cons